Advanced-Recruit-Training/src/LogFileReader.cpp

```diff
@@ -47,25 +48,28 @@ bool LogFileReader::readAll(const QString& fileName)
             emit batteryDataReceived(batteryData);
         }
     }
-
     return true;
 }
 
-/* File input is a csv file in the format of hh:mm:ss:zzz, voltage, current.
- * Negative current values denote a charging battery.
- * Need to implement error checking for the correct number of values and
- * that the conversion from string to double is sucessful.*/
 bool LogFileReader::parseLine(const QString& line, BatteryData& batteryData) const
 {
     QStringList sections = line.split(BATDATA_DELIMITER);
-
+    if(sections.size() != 3 || sections.at(0).isNull() || sections.at(1).isNull() || sections.at(2).isNull())
+    {
+        return false;
+    }
     QString timeString = sections.at(0);
     batteryData.time = QTime::fromString(timeString, STRING_TIME_FORMAT);
+    if(!batteryData.time.isValid())
+        return false;
 
-    batteryData.voltage = sections.at(1).toDouble();
-
-    batteryData.current = sections.at(2).toDouble();
+    bool voltageDoubleFail;
+    bool currentDoubleFail;
+    batteryData.voltage = sections.at(1).toDouble(&voltageDoubleFail);
+    batteryData.current = sections.at(2).toDouble(&currentDoubleFail);
 
+    if(!voltageDoubleFail || !currentDoubleFail)
```

Comment:
This if statement can be combined with the `batteryData.time` check.

Also, wording the if statement as `if(voltageDoubleOk && currentDoubleOk){return true}` instead reduces reviewer overhead; someone reviewing your code wouldn't have to do lots of mental logic to NOT everything in the statement.